emcache/cluster.py

```diff
@@ -71,42 +81,75 @@ def __init__(
         ssl: bool,
         ssl_verify: bool,
         ssl_extra_ca: Optional[str],
+        autodiscovery: bool,
+        autodiscovery_poll_interval: float,
+        autodiscovery_timeout: float,
+        loop: asyncio.AbstractEventLoop,
     ) -> None:
 
         if not memcached_hosts_address:
             raise ValueError("Nodes can't be an empty list, at least one node has to be provided")
 
         self._cluster_events = cluster_events
         self._purge_unhealthy_nodes = purge_unhealthy_nodes
+        self._autodiscovery = autodiscovery
+        self._autodiscovery_poll_interval = autodiscovery_poll_interval
+        self._autodiscovery_timeout = autodiscovery_timeout
+        self._autodiscover_config_version = -1
+        self._first_autodiscovery_done = loop.create_future()
+        self._loop = loop
 
         # Create nodes and configure them to be used by the Rendezvous
         # hashing. By default all are placed under the healthy list, if later
         # on they report that are unhealhty and `purge_unhealthy_nodes` has been
         # configured they will be removed.
-        self._unhealthy_nodes = []
-        self._healthy_nodes = [
-            Node(
-                memcached_host_address,
-                max_connections,
-                min_connections,
-                purge_unused_connections_after,
-                connection_timeout,
-                self._on_node_healthy_status_change_cb,
-                ssl,
-                ssl_verify,
-                ssl_extra_ca,
-            )
-            for memcached_host_address in memcached_hosts_address
+        self._new_node_options = [
+            max_connections,
+            min_connections,
+            purge_unused_connections_after,
+            connection_timeout,
+            self._on_node_healthy_status_change_cb,
+            ssl,
+            ssl_verify,
+            ssl_extra_ca,
         ]
-        self._build_rdz_nodes()
         self._cluster_managment = _ClusterManagment(self)
         self._events = asyncio.Queue(maxsize=MAX_EVENTS)
         self._events_dispatcher_task = asyncio.get_running_loop().create_task(self._events_dispatcher())
         self._closed = False
+
+        if self._autodiscovery:
+            self._unhealthy_nodes = []
+            self._healthy_nodes = []
+            self._discovery_nodes = [
+                Node(
+                    memcached_host_address,
+                    1,
+                    1,
```

Comment:
I did not see this, this definitely will help us to mitigate any issue with having connections opened to nodes that were removed from the cluster but they were used for making the discovery, so worst case scenario we would have one discovery cycle filing and having the TCP recycled, next attempt should work since a new DNS query should be issued which should point to a new IP.